### DISCUSSION TITLE
[QUESTION] Getting tools/preprocess_data.py to work is painful

### DISCUSSION THREAD

**Discussion by sambar1729** [NVIDIA/Megatron-LM#974] [Q&A]
**Your question**
Can `tools/preprocess_data.py` be simplified? 

Using 
```
python tools/preprocess_data.py \
       --input my-corpus.json \
       --output-prefix my-gpt2 \
       --vocab-file gpt2-vocab.json \
       --tokenizer-type GPT2BPETokenizer \
       --merge-file gpt2-merges.txt \
       --append-eod
```

Right now, it requires nltk, torch, transformer_engine, as well as apex. 
Installing transformer_engine does not work out of the box -- had to install out of box (on a A100). 
Installing apex has similar problems, when using https://github.com/NVIDIA/apex?tab=readme-ov-file#linux  

Given that the repo does not have some sample `idx`, `bin` files, one would expect the `preprocess_data` process to be relatively simple. Could this process be simplified?

### Installing apex
```
$ pip install -v --disable-pip-version-check --no-cache-dir --no-build-isolation --config-settings "--build-option=--cpp_ext" --config-settings "--build-option=--cuda_ext" ./
```
gives 
```
....
....
  If this is not desired, please set os.environ['TORCH_CUDA_ARCH_LIST'].
    warnings.warn(
  Emitting ninja build file /home/megauser/apex/build/temp.linux-x86_64-cpython-310/build.ninja...
  Compiling objects...
  Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
  [1/2] c++ -MMD -MF /home/megauser/apex/build/temp.linux-x86_64-cpython-310/csrc/mlp.o.d -pthread -B /home/megauser/.conda/envs/pre/compiler_compat -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/megauser/.conda/envs/pre/include -fPIC -O2 -isystem /home/megauser/.conda/envs/pre/include -fPIC -I/home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include -I/home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/api/include -I/home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/TH -I/home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/THC -I/usr/local/cuda/include -I/home/megauser/.conda/envs/pre/include/python3.10 -c -c /home/megauser/apex/csrc/mlp.cpp -o /home/megauser/apex/build/temp.linux-x86_64-cpython-310/csrc/mlp.o -O3 -DVERSION_GE_1_1 -DVERSION_GE_1_3 -DVERSION_GE_1_5 -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=mlp_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++17
  /home/megauser/apex/csrc/mlp.cpp: In function ‘std::vector<at::Tensor> mlp_forward(int, int, std::vector<at::Tensor>)’:
  /home/megauser/apex/csrc/mlp.cpp:57:21: warning: comparison of integer expressions of different signedness: ‘int’ and ‘long unsigned int’ [-Wsign-compare]
     57 |   for (int i = 0; i < num_layers; i++) {
        |                   ~~^~~~~~~~~~~~
  /home/megauser/apex/csrc/mlp.cpp:64:77: warning: ‘at::DeprecatedTypeProperties& at::Tensor::type() const’ is deprecated: Tensor.type() is deprecated. Instead use Tensor.options(), which in many cases (e.g. in a constructor) is a drop-in replacement. If you were using data from type(), that is now available from Tensor itself, so instead of tensor.type().scalar_type(), use tensor.scalar_type() instead and instead of tensor.type().backend() use tensor.device(). [-Wdeprecated-declarations]
     64 |   auto out = at::empty({batch_size, output_features.back()}, inputs[0].type());
        |                                                                             ^
  In file included from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/ATen/core/Tensor.h:3,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/ATen/Tensor.h:3,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/autograd/function_hook.h:3,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/autograd/cpp_hook.h:2,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/autograd/variable.h:6,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/autograd/autograd.h:3,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/api/include/torch/autograd.h:3,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/csrc/api/include/torch/all.h:7,
                   from /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/torch/extension.h:5,
                   from /home/megauser/apex/csrc/mlp.cpp:1:
  /home/megauser/.conda/envs/pre/lib/python3.10/site-packages/torch/include/ATen/core/TensorBody.h:225:30: note: declared here
    225 |   DeprecatedTypeProperties & type() const {
        |                              ^~~~
````

**Comment by Advait251206** [comment]
This issue is really about **Megatron-LM preprocessing being over-dependant on heavy optional dependencies + noisy C++ extension builds**, not a single bug.

Let’s break it down clearly.

---

# 1. Why `tools/preprocess_data.py` feels painful

You are right: for a script that *only tokenizes text → writes .bin/.idx*, the dependency stack is surprisingly heavy.

### Why it pulls so many dependencies

Even though your command is:

```bash
python tools/preprocess_data.py \
  --input my-corpus.json \
  --output-prefix my-gpt2 \
  --vocab-file gpt2-vocab.json \
  --tokenizer-type GPT2BPETokenizer \
  --merge-file gpt2-merges.txt \
  --append-eod
```

Internally Megatron-LM tries to support:

### A) Multiple tokenizer backends
- GPT2BPETokenizer (HuggingFace-style BPE)
- SentencePieceTokenizer
- HuggingFaceTokenizer (optional newer path)

Some of these paths indirectly pull:
- `transformers`
- `nltk` (for sentence splitting / dataset utilities in older code paths)

---

### B) Optional fused / optimized kernels

These are NOT required for preprocessing, but still get imported in some setups:

- `apex`
- `transformer_engine`
- `torch` extensions

The problem:
> The script is not cleanly isolating "data preprocessing only" vs "training stack"

So your environment ends up installing half the training system just to tokenize text.

---

# 2. About your Apex build warnings

These:

```text
warning: comparison of integer expressions of different signedness
warning: Tensor.type() is deprecated
```

are **NOT errors**.

They come from:

- older Apex CUDA extensions
- modern PyTorch (2.x)
- GCC 11/12 warnings

### Key point:

✔ These warnings do NOT affect correctness  
✔ Build still succeeds  
✔ Apex is just not maintained for modern PyTorch cleanly anymore  

---

# 3. Why Apex / Transformer Engine are not actually needed here

For your command:

```bash
preprocess_data.py
```

👉 Neither of these are logically required:

- ❌ apex (used for fused optimizers, fp16 training)
- ❌ transformer_engine (used for NVIDIA Hopper/A100+ training kernels)

They are only needed for:

- training speedups
- mixed precision kernels
- fused attention / MLP ops

Not dataset preprocessing.

---

# 4. The real design issue in Megatron-LM

This is the core problem:

> preprocessing script imports from the same codebase as training runtime

So dependency leakage happens:

```
preprocess_data.py
   ↓
imports tokenizer
   ↓
imports megatron training utils
   ↓
pulls apex / transformer_engine / torch extensions
```

---

# 5. How people usually fix this

### Option 1 (recommended): minimal install env

Just install:

```bash
pip install torch numpy regex tqdm
```

and patch imports so preprocessing doesn't touch training-only modules.

---

### Option 2: use simpler tokenizer path

Instead of:

```bash
GPT2BPETokenizer
```

use HuggingFace fast tokenizer:

```bash
transformers
```

This avoids some legacy dependencies.

---

### Option 3: disable optional imports (hacky but effective)

In practice people do:

```bash
export NVTE_DISABLE=1
export USE_APEX=0
```

or simply avoid installing apex entirely.

Megatron-LM usually still runs preprocessing fine.

---

# 6. About the compiler warnings in Apex

This part:

```text
Tensor.type() is deprecated
comparison of signed/unsigned
```

is expected because:

- Apex predates PyTorch 2.x API changes
- GCC is stricter now
- code uses older Tensor APIs

You can safely ignore unless build fails.

---

# Final takeaway

### ✔ Your intuition is correct
Yes — preprocessing is unnecessarily heavy.

### ✔ Root cause
Megatron-LM does not separate:
- training runtime dependencies
- preprocessing-only dependencies

### ✔ Practical advice
You do NOT need:
- apex
- transformer_engine
- most CUDA extensions

just to run `preprocess_data.py`.

---

If you want, I can show a **clean minimal Dockerfile for Megatron-LM preprocessing only (no apex, no TE, 10x lighter build)**.

**Comment by Hosseinghorbani0** [comment]
I think there is an important distinction here between the dependencies required by the preprocessing entry point and the dependencies imported indirectly through Megatron-LM's shared modules.

The actual preprocessing path mainly needs to:

1. load the input dataset,
2. tokenize the text,
3. append EOD when requested,
4. and write the `.bin` / `.idx` output.

That makes me wonder whether the cleaner solution would be to make the preprocessing dependency boundary explicit rather than relying on environment-level workarounds.

For example, could the tokenizer and dataset-writing components used by `tools/preprocess_data.py` be isolated into a lightweight module that does not import training-only components such as Apex or Transformer Engine?

This would also make the installation requirements much clearer:

- preprocessing-only users would install only the dependencies needed for tokenization and indexed dataset generation;
- training users could install the optional CUDA/Transformer Engine/Apex components separately.

I think this would be preferable to relying on environment variables or manually patching imports, because it would make `preprocess_data.py` reproducible in a clean environment.

Has there been any consideration of separating the preprocessing dependency graph from the training dependency graph?
